__torch_dispatch__: Populate kwargs dictionary with keyword-only arguments

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62822

This is BC breaking for people who were using the old integration,
although only if you had been writing bindings for functions with
keyword-only arguments (that includes functorch).  Other than that,
the patch was pretty straightforward.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D30134552](https://our.internmc.facebook.com/intern/diff/D30134552)

ezyang committed:
__torch_dispatch__: Populate kwargs dictionary with keyword-only arguments

This is BC breaking for people who were using the old integration,
although only if you had been writing bindings for functions with
keyword-only arguments (that includes functorch).  Other than that,
the patch was pretty straightforward.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30134552).

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1519,18 +1520,28 @@ void concrete_dispatch_fn(const c10::impl::PyInterpreter*, const c10::OperatorHa
-    auto& ivalue = arguments[idx];
-    // Search for Tensors (as they may have the torch functions we need)
+  // Find overloaded tensors
+  for (int64_t idx = 0; idx < kwarg_only_end; idx++) {
```
nit: idx variable re-use for a very different meaning compared to the loops above

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1519,18 +1520,28 @@ void concrete_dispatch_fn(const c10::impl::PyInterpreter*, const c10::OperatorHa
+      break;
+    }
+  }
+  const int64_t kwarg_only_begin = idx + 1;
```
Is there ever a case in the future where these two could be different?

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1548,7 +1559,17 @@ void concrete_dispatch_fn(const c10::impl::PyInterpreter*, const c10::OperatorHa
+  }
+
+  // Populate positional arguments
+  for (int64_t idx = 0; idx < positional_end; idx++) {
```
This is not folded in the loop above for clarity?
Because you could have it there if an `if (idx < positional_end)` and these two cases?

albanD reviewed (commented):
Looks good.
Just some style questions

bdhirsh commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1507,9 +1507,10 @@ void concrete_dispatch_fn(const c10::impl::PyInterpreter*, const c10::OperatorHa
-  // Pre-scan for arguments that match defaults
-  int64_t default_suffix_len = 0;
-  for (int64_t idx = arguments.size() - 1; idx >= 0; idx--) {
+  // Find the first valid idx that isn't defaulted.  E.g., if there
```
Unless i'm misreading - isn't it the last index that isn't defaulted? (or -1 if every arg in the function is defaulted).
e.g.
```
def foo(a, b, c=1, d=1):
    pass
foo(1, 1, 1, 1)
index would be 1 in this case (`b`)

bdhirsh commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1519,18 +1520,28 @@ void concrete_dispatch_fn(const c10::impl::PyInterpreter*, const c10::OperatorHa
+  // (Why +1?  As in the normal indexing convention, end is exclusive, begin is
+  // inclusive.  So if idx is actually kwarg only--we don't know yet--it is
+  // only included if I set end to +1)
+  const int64_t kwarg_only_end = idx + 1;
```
I had to stare at this for a while haha. I think in a case like:
```
def foo(a, b=1, *, c=1, d=1):
    pass
foo(1, 1, 1, 1)
````
It's a little weird because kwarg_only_begin and kwarg_only_end won't actually point to indices for value keyword arguments (they'll both point to the index of `b`), but I guess that doesn't matter, since in any case we don't have to populate any kwargs since they're all defaulted.

ezyang committed:
Update on "__torch_dispatch__: Populate kwargs dictionary with keyword-only arguments"


This is BC breaking for people who were using the old integration,
although only if you had been writing bindings for functions with
keyword-only arguments (that includes functorch).  Other than that,
the patch was pretty straightforward.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D30134552](https://our.internmc.facebook.com/intern/diff/D30134552)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30134552).

ezyang commented:
Based on the multitude of questions on the how exactly the index counting was done, I rewrote the logic for clarity and also to fixed a bug when positional is defaulted but kwarg only isn't.

albanD commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -1507,30 +1507,54 @@ void concrete_dispatch_fn(const c10::impl::PyInterpreter*, const c10::OperatorHa
-    auto& ivalue = arguments[idx];
-    // Search for Tensors (as they may have the torch functions we need)
+  // Find overloaded tensors
+  for (int64_t idx = 0; idx < arguments.size(); idx++) {
```
This used to only go through the non-default arguments, it now goes through all of them in case a kwarg is actually overloaded?

ezyang replied:
Yep. It's harmless semantically, since the default arguments will never be overloaded. It would be a slight perf optimization to skip defaulted arguments but it's a choice between checking "is defaulted" versus checking "is tensor/tensor list" and so it doesn't really matter in the end. (Unlike in the old version, there is no span we can stop, because kwarg and positional are now defaulted/un-defaulted separately)

albanD commented on test/test_python_dispatch.py:
```diff
@@ -138,7 +140,53 @@ def test_out(self) -> None:
+            log_input("z", z)
+            torch.addmv(x, y, z)
+            torch.addmv(x, y, z, beta=2)
+            torch.addmv(x, y, z, alpha=2)
```
do we want a test to make sure that `beta=1` actually gets ignored to make sure that we don't change that behavior in the future by mistake?

ezyang replied:
sure

ezyang committed:
Update on "__torch_dispatch__: Populate kwargs dictionary with keyword-only arguments"


This is BC breaking for people who were using the old integration,
although only if you had been writing bindings for functions with
keyword-only arguments (that includes functorch).  Other than that,
the patch was pretty straightforward.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D30134552](https://our.internmc.facebook.com/intern/diff/D30134552)

[ghstack-poisoned]

ezyang commented:
@ezyang has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30134552).

albanD reviewed (approved):
LGTM
